File: src/main/java/net/lifecity/mc/skillmaster/skill/skills/MoveFast.java
package net.lifecity.mc.skillmaster.skill.skills;

import net.lifecity.mc.skillmaster.SkillMaster;
import net.lifecity.mc.skillmaster.skill.Skill;
import net.lifecity.mc.skillmaster.skill.SkillType;
import net.lifecity.mc.skillmaster.user.SkillUser;
import net.lifecity.mc.skillmaster.weapon.Weapon;
import org.bukkit.Material;
import org.bukkit.Particle;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.util.Vector;

import java.util.Arrays;

public class MoveFast extends Skill {

    public MoveFast(SkillUser user) {
        super(
                "高速移動",
                Arrays.asList(Weapon.STRAIGHT_SWORD, Weapon.DAGGER, Weapon.RAPIER),
                SkillType.MOVE,
                Arrays.asList("向いている方向に高速移動します。", "上方向には飛べません。"),
                0,
                25,
                user
        );
    }

    @Override
    public void activate() {
        super.activate();

        Vector vector = user.getPlayer().getEyeLocation().getDirection().multiply(1.55);

        // Yの値が+だったら
        if (vector.getY() > 0)
            vector.setY(0.15);

        user.getPlayer().setVelocity(vector);

        // 軌道
        new BukkitRunnable() {
            int count = 0;
            @Override
            public void run() {
                if (count >= 7)
                    cancel();
                if (user.getPlayer().getVelocity().length() < 0.3)
                    cancel();
                if (count % 2 == 1)
                    particle(Particle.FALLING_DUST, user.getPlayer().getLocation(), Material.OXIDIZED_COPPER.createBlockData());
                if (count %  2 == 0)
                    particle(Particle.FALLING_DUST, user.getPlayer().getLocation(), Material.IRON_BLOCK.createBlockData());
                if (count % 3 == 0)
                    particle(Particle.FALLING_DUST, user.getPlayer().getLocation(), Material.ICE.createBlockData());
                count++;
            }
        }.runTaskTimer(SkillMaster.instance, 0, 1);
    }
}

File: src/main/java/net/lifecity/mc/skillmaster/skill/skills/MoveFast.kt
package net.lifecity.mc.skillmaster.skill.skills

import com.github.syari.spigot.api.particle.spawnParticle
import com.github.syari.spigot.api.scheduler.runTaskTimer
import net.lifecity.mc.skillmaster.SkillMaster
import net.lifecity.mc.skillmaster.skill.ISkill
import net.lifecity.mc.skillmaster.skill.SkillManager
import net.lifecity.mc.skillmaster.skill.SkillType
import net.lifecity.mc.skillmaster.user.SkillUser
import net.lifecity.mc.skillmaster.weapon.Weapon
import org.bukkit.Material
import org.bukkit.Particle

class MoveFast(
    override val name: String = "高速移動",
    override val weaponList: List<Weapon> = listOf(Weapon.STRAIGHT_SWORD, Weapon.DAGGER, Weapon.RAPIER),
    override val type: SkillType = SkillType.MOVE,
    override val lore: List<String> = listOf("向いている方向に高速移動します。", "上方向には飛べません。"),
    override var isActivated: Boolean = false,
    override val interval: Int = 50,
    override var inInterval: Boolean = false,
    override val user: SkillUser
) : ISkill {

    override fun canActivate(): Boolean = true

    override fun onActivate() {
        user
        val vector = user.player.eyeLocation.direction.multiply(1.55)

        // Yの値が+だったら
        if (vector.y > 0) vector.y = 0.15
        user.player.velocity = vector

        // 軌道
        var count = 0
        SkillMaster.INSTANCE.runTaskTimer(1) {
            if (count >= 7) cancel()
            if (user.player.velocity.length() < 0.3) cancel()

            val loc = user.player.location
            if (count % 2 == 1) loc.spawnParticle(
                Particle.FALLING_DUST,
                data = Material.OXIDIZED_COPPER.createBlockData()
            )
            if (count % 2 == 0) loc.spawnParticle(Particle.FALLING_DUST, data = Material.IRON_BLOCK.createBlockData())
            if (count % 3 == 0) loc.spawnParticle(Particle.FALLING_DUST, data = Material.ICE.createBlockData())
            count++
        }
    }

    override fun onDeactivate() {}
}